src/decoder.rs

```diff
@@ -546,117 +545,143 @@ pub enum ScopeEntry {
     Decoder(Decoder),
 }
 
-pub struct Scope<'a> {
-    parent: Option<&'a Scope<'a>>,
-    names: Vec<Cow<'static, str>>,
-    entries: Vec<ScopeEntry>,
+pub enum Scope<'a> {
+    Empty,
+    Normal(&'a NormalScope<'a>),
+    Single(SingleScope<'a>),
+    Decoder(DecoderScope<'a>),
 }
 
-pub struct ScopeIter<'a> {
-    parent: Option<&'a Scope<'a>>,
-    name_iter: std::iter::Rev<std::slice::Iter<'a, Cow<'static, str>>>,
-    entry_iter: std::iter::Rev<std::slice::Iter<'a, ScopeEntry>>,
+pub struct NormalScope<'a> {
+    parent: &'a Scope<'a>,
+    entries: Vec<(Cow<'static, str>, Value)>,
 }
 
-impl<'a> Iterator for ScopeIter<'a> {
-    type Item = (&'a Cow<'static, str>, &'a ScopeEntry);
-
-    fn next(&mut self) -> Option<Self::Item> {
-        match (self.name_iter.next(), self.entry_iter.next()) {
-            (Some(name), Some(entry)) => Some((name, entry)),
-            _ => match self.parent {
-                Some(parent) => {
-                    *self = parent.into_iter();
-                    self.next()
-                }
-                None => None,
-            },
-        }
-    }
+pub struct SingleScope<'a> {
+    parent: &'a Scope<'a>,
+    name: Cow<'static, str>,
```

Comment:
Is it possible that we can replace this with `&'a str` for cheaper Copy than Clone? We already know that `SingleScope<'a>` is bound by the lifetime constraint `: 'a`, based on the other two elements...

Reply:
If we do this, it might even be possible to make `SingleScope` and `DecoderScope` implement `Copy` as well!

Reply:
I've changed it to `&'a str` although not made it Copy as Scope itself can't be Copy.